keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,1044 @@
+
+#### Story 2
+
+As a platform-optimized CRI runtime developer, I want to know detailed
+container resource requests to be able to make optimal, platform specific,
+resource allocations. Some of the resources may be handled outside cgroups and
+be container runtime specific details.
+
+#### Story 3
+
+As a cluster administrator, I want to install an OCI hook/runc wrapper/NRI
+plugin that does customized resource handling.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+The proposal only adds new informantional data to the CRI API between kubelet
+and the container runtime with no user-visible changes which mitigates possible
+risks considerably.
+
+Data duplication/inconsistency with native resources could be considered a risk
+as those are passed down to CRI both as "raw" requests and limits and as
+"translated" resource control parameters (like cpu shares oom scoring etc). But
+this should be largely mitigated by code reviews and unit tests.
+
+## Design Details
+
+The proposal is that kubelet discloses full resources information from the
+PodSpec to the container runtime. This is accomplished by extending the
+ContainerConfig, UpdateContainerResourcesRequest and PodSandboxConfig messages
+of the CRI API.
+
+With this information, the runtime can for example do detailed resource
+allocation so that CPU, memory and other resources for each container are
+optimally aligned.
+
+The resource information is included in PodSandboxConfig so that the runtime
+can see the full picture of Pod's resource usage at Pod creation time, for
+example enabling more holistic resource allocation and thus better
+interoperability between containers inside the Pod.
+
+### CRI API
+
+#### ContainerConfig
+
+The ContainerConfig message (used in CreateContainer request) is extended to
+contain unmodified resource requests from the PodSpec.
+
+```diff
++import "k8s.io/apimachinery/pkg/api/resource/generated.proto";
+
+ message ContainerConfig {
+ 
+ ...
+ 
+     // Configuration specific to Windows containers.
+     WindowsContainerConfig windows = 16;
+ 
+     // CDI devices for the container.
+     repeated CDIDevice CDI_devices = 17;
++
++    // Kubernetes resource spec of the container
++    KubernetesResources kubernetes_resources = 18;
+ }
+ 
++// KubernetesResources contains the resource requests and limits as specified
```

Comment:
Do you mean including all needed information (mounts, devices, CDI devices) in this single message? Or renaming the `KubernetesResources` here to `RequestedResources`. I think the issue with the former is that mounts, devices and CDI devices are already separate fields in the existing `ContainerConfig` and we can't remove them (and probably not duplicate the same information in a `KubernetesResources` message)

Reply:
ping @elezar I was a bit unsure what you meant. The example yamls under **kubelet** section hopefully now better explains how all informatino is passed down in practice.

Reply:
I was seeing if there was a way to consolidate things so as to not have to make multiple modifications if new resource types / fields are added. Since -- as you point out -- some of the fields of interest are already present in `ContainerConfig`, let's stick with what you have here.